### ISSUE TITLE
Handle partial messages in data channels

### ISSUE THREAD

**Issue by lgrahl** [meetecho/janus-gateway#1091]
I've had a glance at your code and it seems you're not handling the `MSG_EOR` flag which means you will treat incoming partial SCTP messages as whole messages, thus violating the message-oriented principle of SCTP (and data channels). This is a widespread issue in many implementations we're currently trying to clean up.

If you need help, you can look at the [relevant section in RAWRTC's code](https://github.com/rawrtc/rawrtc/blob/master/src/librawrtc/sctp_transport.c#L1406) or at [Mozilla's implementation](https://dxr.mozilla.org/comm-central/source/mozilla/netwerk/sctp/datachannel/DataChannel.cpp#1737). I'm the author of RAWRTC and I've also worked on Mozilla's data channel code, so feel free to ping me if you have any questions.

Tracked (and resolved) for Firefox in [bug 979417](https://bugzilla.mozilla.org/show_bug.cgi?id=979417). For Chromium, this is tracked in [issue 7774](https://bugs.chromium.org/p/webrtc/issues/detail?id=7774).

**Comment by lminiero** [comment]
Thanks for the heads up! I heavily based the SCTP code on one of the examples provided as part of the `usrsctp` repo, so if it wasn't there, it makes sense we're not handling that flag: we definitely didn't add it ourselves, as my knowledge of the library is frankly limited.

What would this entail? Simply fill in a buffer until `MSG_EOR` is set, and only then pass the message to the core in case?

**Comment by lgrahl** [comment]
Yeah, I think many people based it on that example. That would explain why there are so many implementations that don't handle the flag.

> Simply fill in a buffer until `MSG_EOR` is set, and only then pass the message to the core in case?

Exactly that. Be aware that there is a corner case for unreliable messages being aborted. See [my code](https://github.com/rawrtc/rawrtc/blob/master/src/librawrtc/sctp_transport.c#L823) that handles this event (which needs to be registered).

**Comment by lminiero** [comment]
Created a simple PR for the buffering part. I'll work on the corner case you mentioned later on, thanks for the pointer!

**Comment by lminiero** [comment]
As a side note, do you know of any way to test this?

**Comment by lgrahl** [comment]
There isn't an easy way to test this but with my master thesis starting soon, I'll definitely write a test for that case. Ping me in a few months time in case you don't hear from me.

**Comment by lminiero** [comment]
Looking forward to those tests, and best of luck on your master thesis! Closing as I just merged the PR, I'll create a new one later on to address the corner case in the future.

**Comment by lminiero** [comment]
@lgrahl I'm starting to guess that by "handle partial messages" you didn't mention just incoming messages, but also outgoing, right? Or is the delivery automatically frarmented in partial messages when you try to send something that is too large?

**Comment by lgrahl** [comment]
It's only mandatory for incoming messages. If you want to send partial messages, you'll need to explicitly activate that. If you want to do that, I can give you some pointers. Otherwise, you'll be limited by usrsctp's send buffer and will get `EMSGSIZE` if the message is too large. If you haven't changed usrsctp's buffer sizes, this will mean that you can only send messages up to 128 KiB. Anything else will fail and in case you're not emptying the buffer fast enough, other sizes may fail as well.

TL;DR yeah, it completely makes sense to add support for that.
